src/Shared/BufferedReadStream.cs

```diff
@@ -0,0 +1,152 @@
+﻿// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+//-----------------------------------------------------------------------
+// </copyright>
+// <summary>Base class for the implementation of a node endpoint for out-of-proc nodes.</summary>
+//-----------------------------------------------------------------------
+
+using System;
+using System.IO;
+
+namespace Microsoft.Build.BackEnd
+{
+    internal class BufferedReadStream : Stream
+    {
+        const int BUFFER_SIZE = 1024;
+
+        Stream _innerStream;
+        byte[] _buffer;
+
+        int _remainingBufferSize;
+        int _currentIndexInBuffer;
+
+        public BufferedReadStream(Stream innerStream)
+        {
+            _innerStream = innerStream;
+            _buffer = new byte[BUFFER_SIZE];
+
+            _remainingBufferSize = 0;
+        }
+
+        public override bool CanRead { get { return _innerStream.CanRead; } }
+
+        public override bool CanSeek { get { return false; } }
+
+        public override bool CanWrite { get { return _innerStream.CanWrite; } }
+
+        public override long Length { get { return _innerStream.Length; } }
+
+        public override long Position
+        {
+            get { return _innerStream.Position; }
+            set { _innerStream.Position = value; }
```

Comment:
True.  This property shouldn't really be used for this scenario anyway, so I've changed it to throw a `NotSupportedException`.